### DISCUSSION TITLE
Override equality comparision for record struct

### DISCUSSION THREAD

**Discussion by FLAMESpl** [dotnet/csharplang#7018] [Q&A] [answered]
In record classes we can override equality comparision by implementing duck typed `public bool Equals(<type_name>? other)` method.
This approach does not work for structs, unit test in this snippet will fail:

```csharp
public class Test
{
    private class Class
    {
        public int Id { get; }

        public Class(int id)
        {
            Id = id;
        }
    }

    private record struct Record(Class Class)
    {
        public bool Equals(Record? other)
        {
            if (other is null) return false;
            return other.Value.Class.Id == Class.Id;
        }
    }

    [Fact]
    public void MyTest()
    {
        var rec1 = new Record(new Class(0));
        var rec2 = new Record(new Class(0));
        Assert.Equal(rec1, rec2);
    }
}
```

How to override equality comparision for struct records?

**Comment by yaakov-h** [comment, marked as answer]
Take the `?` out of `public bool Equals(Record? other)`. Since this is a `record struct`, `Record?` is `Nullable<Record>` so the method does not have the correct signature to be the correct `Equals(Record)` method.

**Reply by Joe4evr** [reply]
To elaborate: The method signature `public bool Equals(<type_name>? other)` isn't strictly duck-typed - it comes from the fact that `IEquatable<type_name>` is implicitly placed on record declarations. Including the `?` on value types makes your method just another overload instead of matching the `IEquatable` signature which is why it hasn't stopped the compiler from synthesizing its own `Equals` implementation.

**Reply by FLAMESpl** [reply]
ok, thanks it seems that for some reason my Rider IDE generates declaration with question mark

**Reply by yaakov-h** [reply]
sounds like a bug in Rider, it’s probably expecting a record class rather than a record struct.
